File: a7/src/game.js
// ____  ____  ____  __  ____  ____
// / ___)(  _ \(  _ \(  )(_  _)(  __)
// \___ \ ) __/ )   / )(   )(   ) _)
// (____/(__)  (__\_)(__) (__) (____)

class Sprite {
    constructor(type, x, y, w, h, image_url, update_method, onclick_method) {
        this.type = type;
        this.x = x;
        this.y = y;
        this.w = w;
        this.h = h;

        if (image_url != null) {
            this.image = new Image();
            this.image.src = image_url;
            this.image.width = w;
            this.image.height = h;
        }

        this.update = update_method;
        this.onclick = onclick_method;


        // Array of mario images
        if (type = "Mario") {
            this.marioImages = [];
            this.marioImages[0] = new Image();
            this.marioImages[0].src = 'images/mario1.png';

            this.marioImages[1] = new Image();
            this.marioImages[1].src = 'images/mario2.png';

            this.marioImages[2] = new Image();
            this.marioImages[2].src = 'images/mario3.png';

            this.marioImages[3] = new Image();
            this.marioImages[3].src = 'images/mario4.png';

            this.marioImages[4] = new Image();
            this.marioImages[4].src = 'images/mario5.png';
        }

    }

    set_destination(x, y) {
        this.dest_x = x;
        this.dest_y = y;
    }

    ignore_click(x, y) {}

    move(dx, dy) {
        this.dest_x = this.x + dx;
        this.dest_y = this.y + dy;
    }

    marioUpdate() {

        // Update gravity
        if (this.vert_vel === undefined)
            this.vert_vel = 0.0;
        this.vert_vel += 1.2;
        this.y += this.vert_vel;

        if (this.dest_x === undefined)
            return;

        // Move
        if (this.x < this.dest_x)
            this.x;
        else if (this.x > this.dest_x)
            this.x--;
        if (this.y < this.dest_y)
            this.y++;
        else if (this.y > this.dest_y)
            this.y--;

        // Iterate Mario animation
        if (this.marioFrame === undefined)
            this.marioFrame = 0.0;

		if (this.x != this.prev_x)
			this.image = this.marioImages[this.marioFrame++ % 5]

    }

    marioJump() {
        this.vert_vel -= 10;
    }

    brickUpdate() {}
}

// _  _   __  ____  ____  __
// ( \/ ) /  \(    \(  __)(  )
// / \/ \(  O )) D ( ) _) / (_/\
// \_)(_/ \__/(____/(____)\____/

class Model {

    constructor() {
        this.sprites = [];
        //this.sprites.push(new Sprite("lettuce", 50, 200, "images/lettuce.png", Sprite.prototype.brickUpdate, Sprite.prototype.ignore_click));
        this.mario = new Sprite("Mario", 50, 50, 60, 95, "images/mario1.png", Sprite.prototype.marioUpdate, Sprite.prototype.set_destination);
        this.brick = new Sprite("Brick", 20, 400, 1000, 50, null, Sprite.prototype.brickUpdate, Sprite.prototype.ignore_click);
        this.sprites.push(this.mario);
        this.sprites.push(this.brick);
    }

    update() {
        for (let i = 0; i < this.sprites.length; i++) {
            this.sprites[i].update();

            if (this.sprites[i].type != "Mario") {
                if (this.isColliding(this.mario, this.sprites[i]))
                    this.getOut(this.mario, this.sprites[i]);
            }
        }

		if (this.marioFrame === undefined)
            this.marioFrame = 0.0;
		this.scrollPos = this.mario.x - 200;
    }

    onclick(x, y) {
        for (let i = 0; i < this.sprites.length; i++) {
            this.sprites[i].onclick(x, y);
        }
    }

    move(dx, dy) {
        this.mario.move(dx, dy);
    }

    goRight() {
        this.mario.x += 10;
    }

    goLeft() {
        this.mario.x -= 10;
    }

    rememberPreviousPosition() {
        this.mario.prev_x = this.mario.x;
        this.mario.prev_y = this.mario.y;
    }

    isColliding(mario, sprite) {
        if ((mario.x + mario.w) <= sprite.x) {
            //System.out.println("Coming from right");
            return false;
        } else if (mario.x >= (sprite.x + sprite.w)) {
            //System.out.println("Coming from left");
            return false;
        } else if ((mario.y + mario.h) <= sprite.y) {
            //System.out.println("Coming from top");
            // Assume down is positive
            return false;
        } else if (mario.y >= (sprite.y + sprite.h)) {
            //System.out.println("Coming from bottom");
            // Assume down is positive
            return false;
        } else {
            //System.out.println("Colliding with object");
            return true;
        }
    }

    getOut(mario, sprite) {


        // M left side hits B right side
        if (mario.x <= (sprite.x + sprite.w) && mario.prev_x > (sprite.x + sprite.w)) {
            //System.out.println("Hitting Right");
            this.mario.x += 10;
            return;
        }

        // M right side hits B left side
        else if ((mario.x + mario.w) >= sprite.x && (mario.prev_x + mario.w) < sprite.x) {
            //System.out.println("Hitting Left");
            this.mario.x += -10;
            return;
        }

        // M top hits B bottom
        else if (mario.y <= (sprite.y + sprite.h) && mario.prev_y >= (sprite.y + sprite.h)) {
            //System.out.println("Hitting Bottom");
            this.mario.y = b.y + b.h + 1;
            this.mario.vert_vel = 0.0;
            //a.onObject = true;

            // if (b.am_I_a_CoinBlock()) {
            //     Coin c = new Coin(b.x, b.y);
            //     b.hittingBottom = true;
            //     if (b.coinCounter < 5)
            //         soundEffects.playCoinSound();
            // }

            return;
        }

        // M bottom hits B top
        else if ((mario.y + mario.h) >= sprite.y && (mario.prev_y + mario.h) >= sprite.y) {
            //System.out.println("Hitting Top");
            this.mario.y = sprite.y - this.mario.h + 1; // y + h = _y
            this.mario.vert_vel = 3.0;
            //this.mario.jumpCounter = 0;
        }

    }

}




// _  _  __  ____  _  _
// / )( \(  )(  __)/ )( \
// \ \/ / )(  ) _) \ /\ /
// \__/ (__)(____)(_/\_)

class View {
    constructor(model) {
        this.model = model;
        this.canvas = document.getElementById("myCanvas");
        this.turtle = new Image();
        this.turtle.src = "images/turtle.png";
    }

    update() {
        let ctx = this.canvas.getContext("2d");
        ctx.clearRect(0, 0, 1000, 500);
        for (let i = 0; i < this.model.sprites.length; i++) {
            let sprite = this.model.sprites[i];

			if (sprite.type == "Mario") {

			}
			else if (sprite.type == "Brick") {
                ctx.fillStyle = "#a04416";
                ctx.fillRect(sprite.x, sprite.y, sprite.w, sprite.h);
                ctx.stroke();
            } else
                ctx.drawImage(sprite.image, sprite.x, sprite.y);

        }
    }
}

// ___  __   __ _  ____  ____   __   __    __    ____  ____
// / __)/  \ (  ( \(_  _)(  _ \ /  \ (  )  (  )  (  __)(  _ \
// ( (__(  O )/    /  )(   )   /(  O )/ (_/\/ (_/\ ) _)  )   /
// \___)\__/ \_)__) (__) (__\_) \__/ \____/\____/(____)(__\_)

class Controller {
    constructor(model, view) {
        this.model = model;
        this.view = view;
        this.key_right = false;
        this.key_left = false;
        this.key_up = false;
        this.key_down = false;
        let self = this;
        view.canvas.addEventListener("click", function(event) {
            self.onClick(event);
        });
        document.addEventListener('keydown', function(event) {
            self.keyDown(event);
        }, false);
        document.addEventListener('keyup', function(event) {
            self.keyUp(event);
        }, false);
    }

    onClick(event) {
        this.model.onclick(event.pageX - this.view.canvas.offsetLeft, event.pageY - this.view.canvas.offsetTop);
    }

    keyDown(event) {
        if (event.keyCode == 39) this.key_right = true;
        else if (event.keyCode == 37) this.key_left = true;
        else if (event.keyCode == 38) this.key_up = true;
        else if (event.keyCode == 40) this.key_down = true;
        else if (event.keyCode == 32) this.key_space = true;

    }

    keyUp(event) {
        if (event.keyCode == 39) this.key_right = false;
        else if (event.keyCode == 37) this.key_left = false;
        else if (event.keyCode == 38) this.key_up = false;
        else if (event.keyCode == 40) this.key_down = false;
        else if (event.keyCode == 32) this.key_space = false;
    }

    update() {
        this.model.rememberPreviousPosition();

        let dx = 0;
        let dy = 0;

        if (this.key_right)
            this.model.goRight();
        if (this.key_left)
            this.model.goLeft();
        if (this.key_up || this.key_space)
            this.model.mario.marioJump();

        if (dx != 0 || dy != 0)
            this.model.move(dx, dy);
    }
}

// ___   __   _  _  ____
// / __) / _\ ( \/ )(  __)
// ( (_ \/    \/ \/ \ ) _)
// \___/\_/\_/\_)(_/(____)

class Game {
    constructor() {
        this.model = new Model();
        this.view = new View(this.model);
        this.controller = new Controller(this.model, this.view);
    }

    onTimer() {
        this.controller.update();
        this.model.update();
        this.view.update();
    }
}


let game = new Game();
let timer = setInterval(function() {
    game.onTimer();
}, 40);

File: a9/src/game.py
# Andrew Cash
# Programming Paradigms
# Fall 2018
# Assignment 9

import pygame
import time
import random

from pygame.locals import*
from time import sleep


#   _____ ____  ____   ____  ______    ___
#  / ___/|    \|    \ l    j|      T  /  _]
# (   \_ |  o  )  D  ) |  T |      | /  [_
#  \__  T|   _/|    /  |  | l_j  l_jY    _]
#  /  \ ||  |  |    \  |  |   |  |  |   [_
#  \    ||  |  |  .  Y j  l   |  |  |     T
#   \___jl__j  l__j\_j|____j  l__j  l_____j



class Sprite():
    def __init__(self, _type, _x, _y, _w, _h, update_method):
        self.type = _type
        self.x = _x
        self.y = _y
        self.w = _w
        self.h = _h
        self.vert_vel = 0
        self.marioFrame = 0
        self.coinCounter = 0
        self.hittingBottom = False

        self.update = update_method

        if (self.type == "Coin"):
            self.horz_vel = random.randint(-10, 10)
            self.vert_vel = random.randint(0, 5)

    def marioUpdate(self):
        #print("Mario Update!")

        # Update gravity
        self.vert_vel += 1.2
        self.y += self.vert_vel

        # Update animation
        if (self.x != self.prev_x):
            self.marioFrame = ((self.marioFrame + 1) % 5)


    def jump(self):
        #print("Jump")
        self.jumpCounter += 1
        if (self.jumpCounter < 5):
            self.vert_vel -= 5

    def goRight(self):
        self.x += 10

    def goLeft(self):
        self.x -= 10

    def brickUpdate(self):
        #print("Brick Update")
        pass

    def coinBlockUpdate(self):
        pass

    def coinUpdate(self):
        self.vert_vel += 1.2
        self.y += self.vert_vel
        self.x += self.horz_vel



#  ___ ___   ___   ___      ___  _
# |   T   T /   \ |   \    /  _]| T
# | _   _ |Y     Y|    \  /  [_ | |
# |  \_/  ||  O  ||  D  YY    _]| l___
# |   |   ||     ||     ||   [_ |     T
# |   |   |l     !|     ||     T|     |
# l___j___j \___/ l_____jl_____jl_____j


class Model():
    def __init__(self):
        self.sprites = []

        self.mario = Sprite("Mario", 50, 50, 60, 95, Sprite.marioUpdate)
        self.brick0 = Sprite("Brick", 20, 400, 1000, 50, Sprite.brickUpdate)
        self.brick1 = Sprite("Brick", 600, 350, 50, 50, Sprite.brickUpdate)
        self.coinBlock = Sprite("CoinBlock", 250, 100, 89, 83, Sprite.coinBlockUpdate)
        self.sprites.append(self.mario)
        self.sprites.append(self.brick0)
        self.sprites.append(self.brick1)
        self.sprites.append(self.coinBlock)

        self.dest_x = 0
        self.dest_y = 0

        self.scrollPos = 0


    def update(self):
        for i in self.sprites:
            #print(i.type)
            i.update(i)
            if (i.type != "Mario"):
                if (i.type == "Coin"):
                    pass
                if (self.isColliding(self.mario, i)):
                    self.getOut(self.sprites[0], i)
            if (i.type == "Mario"): # fyi this is here becasue mario doesnt have access tothe model and marioRect is stored in the model
                self.marioRect.left = i.x - self.scrollPos
                self.marioRect.top = i.y
            if (i.type == "CoinBlock"):
                self.coinBlockRect.left = i.x - self.scrollPos
                self.coinBlockRect.top = i.y
            if (i.type == "Coin"):
                self.coinRect.left = i.x - self.scrollPos
                self.coinRect.top = i.y

            self.backgroundRect.left = -(self.scrollPos / 2)


        self.scrollPos = self.mario.x - 20

    def set_dest(self, pos):
        self.dest_x = pos[0]
        self.dest_y = pos[1]

    def rememberPreviousPosition(self):
        self.mario.prev_x = self.mario.x
        self.mario.prev_y = self.mario.y

    def isColliding(self, mario, sprite):
        if (sprite.type == "Coin"):
            return

        if ((mario.x + mario.w) <= sprite.x):
            # print("Coming from right")
            return False
        elif (mario.x >= (sprite.x + sprite.w)):
            # print("Coming from left");
            return False
        elif((mario.y + mario.h) <= sprite.y):
            # print("Coming from top");
            # Assume down is positive
            return False
        elif(mario.y >= (sprite.y + sprite.h)):
            # print("Coming from bottom");
            # Assume down is positive
            return False
        else:
            # print("Colliding with object");
            return True

    def getOut(self, mario, sprite):
        # M left side hits B right side
        if (mario.x <= (sprite.x + sprite.w) and mario.prev_x > (sprite.x + sprite.w)):
            # print("Hitting Right")
            self.mario.x += 10
            return

        # M right side hits B left side
        elif ((mario.x + mario.w) >= sprite.x and (mario.prev_x + mario.w) < sprite.x):
            # print("Hitting Left")
            self.mario.x += -10
            return

        # M top hits B bottom
        elif (mario.y <= (sprite.y + sprite.h) and mario.prev_y >= (sprite.y + sprite.h)):
            # print("Hitting Bottom")
            self.mario.y = sprite.y + sprite.h + 1
            self.mario.vert_vel = 1
            # a.onObject = true;

            if (sprite.type == "CoinBlock"):
                sprite.hittingBottom = True

                if (sprite.coinCounter < 5):
                    s = Sprite("Coin", (sprite.x + (sprite.w / 2)), sprite.y, 75, 75, Sprite.coinUpdate)
                    self.sprites.append(s)
                    sprite.coinCounter += 1
                    sprite.hittingBottom = False
                    print(sprite.coinCounter)
            return

        # M bottom hits B top
        elif ((mario.y + mario.h) >= sprite.y and (mario.prev_y + mario.h) >= sprite.y):
            #print("Hitting Top")
            self.mario.y = sprite.y - self.mario.h + 1  # y + h = _y
            self.mario.vert_vel = 0
            self.mario.jumpCounter = 0
            return


#  __ __  ____    ___ __    __
# |  T  |l    j  /  _]  T__T  T
# |  |  | |  T  /  [_|  |  |  |
# |  |  | |  | Y    _]  |  |  |
# l  :  ! |  | |   [_l  `  '  !
#  \   /  j  l |     T\      /
#   \_/  |____jl_____j \_/\_/



class View():
    def __init__(self, model):
        screen_size = (800, 600)
        self.screen = pygame.display.set_mode(screen_size, 32)
        self.model = model

        self.background_image = pygame.image.load("images/background.jpg")
        self.mario_image = pygame.image.load("images/mario1.png")
        self.coinBlock_image = pygame.image.load("images/block1.png")
        self.coinBlockEmpty_image = pygame.image.load("images/block2.png")
        self.coin_image = pygame.image.load("images/coin.png")

        self.model.backgroundRect = self.background_image.get_rect()
        self.model.marioRect = self.mario_image.get_rect()
        self.model.coinBlockRect = self.coinBlock_image.get_rect()
        self.model.coinRect = self.coin_image.get_rect()

    def update(self):
        #print (self.model.mario.marioFrame)

        self.screen.fill([0, 200, 100])

        self.screen.blit(self.background_image, self.model.backgroundRect)

        for i in self.model.sprites:
            if i.type == "Brick":
                brickColor = (160, 68, 22)
                pygame.draw.rect(self.screen, brickColor, (i.x - self.model.scrollPos, i.y, i.w, i.h), 0)
            elif i.type == "Mario":
                self.screen.blit(self.mario_image, self.model.marioRect)
            elif i.type == "CoinBlock":
                if (i.coinCounter < 5):
                    self.screen.blit(self.coinBlock_image, self.model.coinBlockRect)
                else:
                    self.screen.blit(self.coinBlockEmpty_image, self.model.coinBlockRect)
            elif i.type == "Coin":
                self.screen.blit(self.coin_image, self.model.coinRect)

        # Update the full display Surface to the screen
        pygame.display.flip()

#     __   ___   ____   ______  ____    ___   _      _        ___  ____
#    /  ] /   \ |    \ |      T|    \  /   \ | T    | T      /  _]|    \
#   /  / Y     Y|  _  Y|      ||  D  )Y     Y| |    | |     /  [_ |  D  )
#  /  /  |  O  ||  |  |l_j  l_j|    / |  O  || l___ | l___ Y    _]|    /
# /   \_ |     ||  |  |  |  |  |    \ |     ||     T|     T|   [_ |    \
# \     |l     !|  |  |  |  |  |  .  Yl     !|     ||     ||     T|  .  Y
#  \____j \___/ l__j__j  l__j  l__j\_j \___/ l_____jl_____jl_____jl__j\_j


class Controller():
    def __init__(self, model):
        self.model = model
        self.keep_going = True

    def update(self):
        self.model.rememberPreviousPosition()

        for event in pygame.event.get():
            if event.type == QUIT:
                self.keep_going = False
            elif event.type == KEYDOWN:
                if event.key == K_ESCAPE:
                    self.keep_going = False
            elif event.type == pygame.MOUSEBUTTONUP:
                self.model.set_dest(pygame.mouse.get_pos())
        keys = pygame.key.get_pressed()
        if keys[K_LEFT]:
            self.model.sprites[0].goLeft()
        if keys[K_RIGHT]:
            self.model.sprites[0].goRight()
        if keys[K_UP] or keys[K_SPACE]:
            self.model.sprites[0].jump()

#  ___ ___   ____  ____  ____
# |   T   T /    Tl    j|    \
# | _   _ |Y  o  | |  T |  _  Y
# |  \_/  ||     | |  | |  |  |
# |   |   ||  _  | |  | |  |  |
# |   |   ||  |  | j  l |  |  |
# l___j___jl__j__j|____jl__j__j

print("Use the arrow keys to move. Press Esc to quit.")
pygame.init()
m = Model()
v = View(m)
c = Controller(m)
while c.keep_going:
    c.update()
    m.update()
    v.update()
    sleep(0.04)
print("Goodbye")


# ASCII art from http://patorjk.com/software/taag/#p=display&f=Crawford&t=Sprite
